book/src/dev/rfcs/0012-value-pools.md

```diff
@@ -0,0 +1,437 @@
+impl Default for ValueBalance<C> {
+    
+}
+```
+
+### Create a method in `Transaction` that returns `ValueBalance<NegativeAllowed>` for the transaction
+
+We first add `value_balance()` methods in all the modules we need and use them to get the value balance for the whole transaction.
+
+#### Create a method in `Input` that returns `ValueBalance<NegativeAllowed>`
+
+- Method location is at `zebra-chain/src/transparent.rs`.
+- Method need `utxos`, this information is available in `verify_transparent_inputs_and_outputs`.
+- If the utxos are not available in the block or state, verification will timeout and return an error
+
+```rust
+impl Input {
+    fn value_balance(&self, utxos: &HashMap<OutPoint, Utxo>) -> ValueBalance<NegativeAllowed> {
+
+    }
+}
+```
+
+#### Create a method in `Output` that returns `ValueBalance<NegativeAllowed>`
+
+- Method location is at `zebra-chain/src/transparent.rs`.
+
+```rust
+impl Output {
+    fn value_balance(&self) -> ValueBalance<NegativeAllowed> {
+
+    }
+}
+```
+
+#### Create a method in `JoinSplitData` that returns `ValueBalance<NegativeAllowed>`
+
+- Method location is at `zebra-chain/src/transaction/joinsplit.rs`
+
+```rust
+pub fn value_balance(&self) -> ValueBalance<NegativeAllowed> {
+
+}
+```
+
+#### Create a method in `sapling::ShieldedData` that returns `ValueBalance<NegativeAllowed>`
+
+- Method location is at `zebra-chain/src/transaction/sapling/shielded_data.rs`
+
+```rust
+pub fn value_balance(&self) -> ValueBalance<AllowNegative> {
+
+}
+```
+
+#### Create a method in `orchard::ShieldedData` that returns `ValueBalance<NegativeAllowed>`
+
+- Method location is at `zebra-chain/src/transaction/orchard/shielded_data.rs`
+
+```rust
+pub fn value_balance(&self) -> ValueBalance<AllowNegative> {
+
+}
+```
+
+#### Create the `Transaction` method
+
+- Method location: `zebra-chain/src/transaction.rs`
+
+```rust
+/// utxos must contain the utxos of every input in the transaction
+/// TODO: what about UTXOs created and spent in the transaction?
+pub fn value_balance(&self, utxos: &HashMap<transparent::OutPoint, Utxo>) -> Result<ValueBalance<NegativeAllowed>, Err> {
+    match self {
+        Transaction::V1 { .. } => {},
+        Transaction::V2 { .. } => {},
+        Transaction::V3 { .. } => {},
+        Transaction::V4 { .. } => {},
+        Transaction::V5 { .. } => {
+            
+        },
+    }
+}
+```
+
+### Create a method in `Block` that returns `ValueBalance<NegativeAllowed>` for the block
+
+- Method location is at `zebra-chain/src/block.rs`.
+- Method will make use of `Transaction::value_balance` method created before.
+
+```rust
+/// utxos must contain the utxos of every input in the block
+/// TODO: what about UTXOs created and spent in the block?
+pub fn value_balance(&self, utxos: &HashMap<transparent::OutPoint, Utxo>) -> ValueBalance<NegativeAllowed> {
+    self.transactions()
+        .map(Transaction::value_balance)
+        .sum()
+        .unwrap_or_else(ValueBalance::default)
```

Comment:
@oxarbitrage each block is guaranteed to have at least one coinbase transaction.

Do we want to return an error here instead?
Or do we want to panic?

It seems risky to just return a default when we know the block is invalid.

We'll also have to propagate any errors in the `ValueBalance` sums.

Reply:
Thanks, it seems we should panic. I changed to that at https://github.com/ZcashFoundation/zebra/pull/2430/commits/117a4e8bbde0bb820d6256f6146d3d2b95a06f31